File: ps/0401/1263.cpp
#include <cstdio>
#include <algorithm>
using namespace std;
const int MAX = 1000;
int adj[MAX][MAX];
int main() {
    freopen("1263_input.txt", "r", stdin);
    int T, N, temp;
    scanf("%d", &T);
    for (int tc = 0; tc < T; tc++) {
        scanf("%d", &N); 
        for (int y = 0; y < N; y++)
            for (int x = 0; x < N; x++) {
                scanf("%d", &temp);
                if (y == x) {
                    adj[y][x] = 0;
                } else if (temp) {
                    adj[y][x] = temp;
                } else {
                    adj[y][x] = 10000;
                }
            }

        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++) {

                if (i == j) continue;
                for (int k = 0; k < N; k++)
                    adj[i][j] = min(adj[i][j], adj[i][k] + adj[j][k]);
            }

        int result = 2147483647;
        for (int i = 0; i < N; i++) {
            temp = 0;
            for (int j = 0; j < N; j++) {
                temp += adj[i][j];
            }
            result = min(result, temp);
        }
        printf("#%d %d\n", tc+1, result);
    }

    return 0;
}

File: ps/0401/1263.py
import sys
sys.stdin = open("1263_input.txt")
sys.stdout = open("1263_output.txt", "w")
T = int(input())
for tc in range(T):
    raw = list(map(int, input().split()))
    N = raw[0]
    adj = []
    a = 1
    distance = [[0 for _ in range(N)] for _ in range(N)]
    for _ in range(N):
        adj.append(raw[a:a+N])
        a += N
    for i in range(N):
        for j in range(N):
            if i == j:
                distance[i][j] = 0
            elif adj[i][j]:
                distance[i][j] = adj[i][j]
            else:
                distance[i][j] = 9999
    for i in range(N):
        for j in range(N):
            for k in range(N):
                distance[i][j] = min(distance[i][j], distance[i][k] + distance[k][j])
    print(f"#{tc+1}", min(sum(row) for row in distance))
